[FABCE-149] Receipt logs should default to empty array

abrinckm opened the pull request:
Resolves issue [FABCE-149](https://jira.hyperledger.org/projects/FABCE/issues/FABCE-149?filter=allopenissues)

MHBauer commented:
I see three issues

1. gofmt the change
1. fix the unit tests
1. the integration tests aren't going to work until we make a change or update our base to 2.0 (or maybe 1.4.6)

MHBauer commented:
And thanks for the contribution. It may take us a bit to fully evaluate and merge it.

abrinckm commented:
Okay. I'll work on unit tests and formatting

abrinckm commented:
Commenting to keep the PR fresh (noticed it went stale). Unit tests fixed and formatting fixed.

MHBauer commented:
Thanks again for the patch. The update looks good for the two mentioned issues.
We're going to need this rebased on master to pick up a9d103e9f8c7353c84ba6f26862f16fccb717985 and get the results of the integration tests. After which there is probably going to be another update needed to fix any test failures there.
Thanks for your patience.

abrinckm commented:
Excellent. Okay I will work on making this an output marshaller. Thank you!

MHBauer commented:
It would require a new test in that package.

Something like https://github.com/hyperledger/fabric-chaincode-evm/blob/2b037c904e99c7b965f166259c62cdbeaadfdf28/fab3/types/types_test.go#L187-L214 would correspond to the marshaller previously discussed..

abrinckm commented:
@MHBauer 
Hi, 

Sorry, I'm wondering if you can answer a question I have.

First, I have defined the marshaller as you've suggested, and written a test, but now I'm confused as to where or how these marshallers are being used in code. I've done a grep on the codebase and I can only find them called in their corresponding test file `types_test.go`

Can you show me as an example how Transaction.MarshalJSON() is being used? Or maybe this is something I don't need to worry about?

Thanks!

abrinckm commented:
Actually nvm my last question. I've learned that they are automatically called whenever their structs are marshalled into json. I'll push my changes shortly.

MHBauer commented on fab3/types/types.go:
```diff
@@ -200,6 +200,17 @@ type Log struct {
+func (txReceipt *TxReceipt) MarshalJSON() ([]byte, error) {
+	receipt := *txReceipt
+
+	if receipt.Logs == nil || len(receipt.Logs) == 0 {
+		receipt.Logs = []Log{}
+	}
```
Sorry, what does the check for 0 length help us with?

abrinckm replied:
Was making sure all bases are covered, but thinking about it now, this may not be needed actually.

MHBauer replied:
My analysis is if we have the nil slice come in, it will fail both of those checks, so the only check that would catch the second clause (and make some difference in output) is for an empty slice, that has more than 0 capacity. Even so, it wouldn't change the resulting output.

I think we can drop the length check.

MHBauer replied:
My analysis is if we have the nil slice come in, it will fail both of those checks, so the only check that would catch the second clause (and make some difference in output) is for an empty slice, that has more than 0 capacity. Even so, it wouldn't change the resulting output.

I think we can drop the length check.

MHBauer commented:
If we want to ensure that this particular issue does not reoccur, we should also add to the integration test https://github.com/hyperledger/fabric-chaincode-evm/blob/master/integration/fab3/fab3_test.go#L123-L139 a check for an empty array at `receipt.Logs`, right?

swetharepakula commented:
One other change I would recommend is updating the documentation so it corresponds with your change: https://github.com/hyperledger/fabric-chaincode-evm/blob/master/Fab3_Instructions.md#eth_getTransactionReceipt.

MHBauer commented:
Agree, specifically to add an example of the logs entry, or the empty array value, https://github.com/hyperledger/fabric-chaincode-evm/blame/master/Fab3_Instructions.md#L297

abrinckm commented:
The integration tests _were_ working after I rebased, but are now failing again after my last commit. 

I'll roll back my commit and see if that helps. 

Also, I'll add the integration test and documentation as suggested.

abrinckm commented:
Okay. Looks like my latest change broke integration tests. Unfortunately I'm unable to run integration tests locally. I'm running into this issue: https://jira.hyperledger.org/browse/FAB-15665

I'll work through that and then revisit here once I have things working again.

MHBauer commented:
@abrinckm if you're on a mac, I have a patch for that. https://github.com/MHBauer/fabric/commit/59a54bea995fad11090ce8b323bbcc215aec0905

abrinckm commented:
> @abrinckm if you're on a mac, I have a patch for that. [MHBauer/fabric@59a54be](https://github.com/MHBauer/fabric/commit/59a54bea995fad11090ce8b323bbcc215aec0905)

that worked

abrinckm commented:
@MHBauer @swetharepakula 

Alright, I think it's good for another review. I modified the integration test so we don't see a reoccurrence and changed the examples to reflect logs as an empty array.

Reverting the unit tests causes them to fail. I believe those changes are still required for the tests to pass even with the new output marshaller. This is because the receipt structures we compare against are simply constructed inline and never marshalled. So the constructor needs the logs field set for it to be correct.

swetharepakula commented on integration/fab3/fab3_test.go:
```diff
@@ -131,6 +131,7 @@ var _ = Describe("Fab3", func() {
 
 		receipt := rpcResp.Result
 
+		Expect(receipt.Logs).To(Equal([]types.Log{}))
```
Use a gomega matcher here instead. Though this does test the change you made, I think the integration test change should be in the web3 specific tests: https://github.com/hyperledger/fabric-chaincode-evm/blob/master/integration/fab3/web3_e2e_test.js#L71-L73. What do you think @MHBauer ?
```suggestion
		Expect(receipt.Logs).ToNot(BeNil())
```

MHBauer replied:
Suggested change looks good.

I don't see what web3 has to do with it.

abrinckm commented:
@MHBauer @swetharepakula No worries! I'm determined to get this fixed. 

I've tried your suggestion. I've reverted my changes in the ethservice.go, but the receipt logs are returned as nil which reintroduces the original bug.

MHBauer commented:
Made an example branch and pr on my fork to check the outupt 
https://github.com/MHBauer/fabric-chaincode-evm/pull/4
build it
https://dev.azure.com/mbauer0254/fabric/_build/results?buildId=18&view=results

MHBauer commented:
As an example, if I turn on verbose test output
```
{"jsonrpc":"2.0","id":16,"method":"eth_getTransactionReceipt","params":["98060ced4d3faf8666d38538400e1c7e002ba90c10233390184d9599c857f57f"]}
[e][Org1.peer0] 2020-04-15 08:52:23.241 PDT [endorser] callChaincode -> INFO 08a [testchannel][a1f602bf] Entry chaincode: name:"qscc" 
[e][Org1.peer0] 2020-04-15 08:52:23.241 PDT [endorser] callChaincode -> INFO 08b [testchannel][a1f602bf] Exit chaincode: name:"qscc"  (0ms)
[e][Org1.peer0] 2020-04-15 08:52:23.241 PDT [comm.grpc.server] 1 -> INFO 08c unary call completed {"grpc.start_time": "2020-04-15T08:52:23.241-07:00", "grpc.service": "protos.Endorser", "grpc.method": "ProcessProposal", "grpc.request_deadline": "2020-04-15T08:55:23.241-07:00", "grpc.peer_address": "127.0.0.1:33336", "grpc.code": "OK", "grpc.call_duration": "800.188µs"}
{"jsonrpc":"2.0","result":{"transactionHash":"0x98060ced4d3faf8666d38538400e1c7e002ba90c10233390184d9599c857f57f","transactionIndex":"0x0","blockHash":"0x62167e704c96f8ff19556c72b8ffaa79ffd3d4ffd30e1d195d012c67dd7a3972","blockNumber":"0x5","contractAddress":"0xd0fe8897d3ffb144a374788dd20e0048dd8dc956","gasUsed":0,"cumulativeGasUsed":0,"to":"","logs":[],"status":"0x1","from":"0xc2c468b207d68e0ca079e45eb3eefef2ecb0810d"},"id":16}
```
in the last line, I can see raw json is `,"logs":[],`

MHBauer commented:
Yup, squash the last two and drop the first two, and this is merged.

MHBauer commented:
Not sure why checks are yellow, but I would ignore them and do the update.

abrinckm commented:
@MHBauer Okay done!

abrinckm commented:
@MHBauer @swetharepakula  Thanks for your patience!

swetharepakula commented:
Thanks @abrinckm! I hope you continue contributing to our project!
